### PR TITLE
Add draft of ISOLATE_STDLIB_MODULES.rst

### PR DESCRIPTION
**Pull request by encukou**
@erlend-aasland, here's a rough draft of a PEP.

### CODE REVIEW

File: ISOLATE_STDLIB_MODULES.rst [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+converted modules are checked and fixed.
+
+Static types that do not need module state access and have no other reason to
+be converted should stay static.
+
+
+Testing
+-------
+
+XXX How should this be tested?
```
Comment (erlend-aasland) [inline ISOLATE_STDLIB_MODULES.rst]: We've already described how to test for disallowed instantiation and serialisation, so how about moving those advices down here? We should also include a short example of how to test for immutability (try to create a new attribute on the type; make sure `TypeError` is raised).

File: ISOLATE_STDLIB_MODULES.rst [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+
+
+Reference Implementation
+========================
+
+Main branch, particularly commits for these issues:
+
+- https://bugs.python.org/issue40077
+- https://bugs.python.org/issue46417
+- https://bugs.python.org/issue1635741
```
Comment (erlend-aasland) [inline ISOLATE_STDLIB_MODULES.rst]: I think we should add links to the actual implementation instead (and maybe one or two reference PRs); asking someone to extract relevant bits and pieces from bpo-1635741 is ... not nice :)

File: ISOLATE_STDLIB_MODULES.rst [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+
+
+Rejected Ideas
+==============
+
+Keep things as they were and not break anything
```
Comment (erlend-aasland) [inline ISOLATE_STDLIB_MODULES.rst]: Would not this section make more sense in the subinterpreters or per-interpreter GIL PEPs? It feels kinda out of place here. Also, if we are to keep it, I would rephrase it to something more accurate and direct, such as _"Do not isolate modules in the stdlib"_.

It is not clear to me what "things" is. I would guess that the sentence implies _"don't convert static types to heat types"_, _"don't convert global state to module state"_, _"don't apply PEP 630 to the stdlib"_, or maybe _"drop subinterpreter support"_. It is hard to tell. Currently, I'm afraid this will only be a source of confusion.

**Review by erlend-aasland — COMMENTED** [review]
This is a very nice draft; thanks! I left some comments after reading through it once. I'll have another round later today or tomorrow.

File: ISOLATE_STDLIB_MODULES.rst (hunk 2 above)
Comment (erlend-aasland) [reply ISOLATE_STDLIB_MODULES.rst]: The `csv` module should be a good candidate.

File: ISOLATE_STDLIB_MODULES.rst [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+
+We submit this PEP to explain the changes, seek consensus on
+whether they are good,  propose changing modules that do not exhibit
+issues, and set best practices for new modules.
+
+We are aware that the pre-existing changes might need to be reverted if issues
+are found.
+
+The PEP takes the form of the proposal that should have been submitted,
+talking about the situation before the proposed changes as current.
```
Comment (erlend-aasland) [inline ISOLATE_STDLIB_MODULES.rst]: We should consider removing this sentence or rewording it so it is easier to grasp.

File: ISOLATE_STDLIB_MODULES.rst [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+
+Note on Backdating
+==================
+
+Much of this proposal has already been implemented, partly to solve related
+issues and partly because of misunderstandings and lack of coordination.
+
+We submit this PEP to explain the changes, seek consensus on
+whether they are good,  propose changing modules that do not exhibit
+issues, and set best practices for new modules.
```
Comment (erlend-aasland) [inline ISOLATE_STDLIB_MODULES.rst]: _[...] propose changing modules that do not exhibit issues [...]_

This reads a little bit weird; did you mean "propose _not_ changing modules that do not exhibit issues"? In that case, we should consider rewording it.

File: ISOLATE_STDLIB_MODULES.rst [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+  For example, ``_csv.field_size_limit`` will get/set a module-specific
+  number. On the other hand, functions like ``readline.get_history_item`` or
+  ``os.getpid`` will continue to work with state that is process-global
+  (external to the module, and possibly shared across other libraries, including
+  non-Python ones).
+
+Conversion to heap types
+------------------------
+
+Types whose methods need access to “their” module instance will be converted
```
Comment (erlend-aasland) [inline ISOLATE_STDLIB_MODULES.rst]:
```suggestion
Types whose methods need access to their module instance will be converted
```

File: ISOLATE_STDLIB_MODULES.rst [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+- All heap types **must fully implement the GC protocol**. See
+  [bpo-42972](https://bugs.python.org/issue42972).
+  (XXX move this to PEP 630)
+
+- All standard library types that used to be static types should remain
+  immutable. Heap types must be defined with the `Py_TPFLAGS_IMMUTABLE_TYPE`
+  flag to retain immutability.
+  See `bpo-43908 <https://bugs.python.org/issue43908>`__.
+
+- A static type with ``tp_new = NULL`` does not have public constructor, but
```
Comment (erlend-aasland) [inline ISOLATE_STDLIB_MODULES.rst]:
```suggestion
- A static type with ``tp_new = NULL`` does not have a public constructor, but
```

File: ISOLATE_STDLIB_MODULES.rst [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+  for an example.
+
+- Use strong "back-refs" to the module object to ensure the module state
+  pointer never outlives objects that access module state. Keep this in mind
+  for external library callbacks that access module state.
+  (XXX move this to PEP 630)
+
+These issues will be added to the Devguide to help any future conversions.
+
+If another kind of issue is found, the module in questions should be unchanged
```
Comment (erlend-aasland) [inline ISOLATE_STDLIB_MODULES.rst]:
```suggestion
If another kind of issue is found, the module in question should be unchanged
```

File: ISOLATE_STDLIB_MODULES.rst [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+until a solution is found and added to the Devguide, and already
+converted modules are checked and fixed.
+
+Static types that do not need module state access and have no other reason to
+be converted should stay static.
```
Comment (erlend-aasland) [inline ISOLATE_STDLIB_MODULES.rst]:
```suggestion
Static types that do not need module state access, and have no other reason to
be converted, should stay static.
```

File: ISOLATE_STDLIB_MODULES.rst [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+-------
+
+XXX How should this be tested?
+
+
+Process
+-------
+
+The following process should be added to the Devguide, and stay there until
+all modules are converted.
```
Comment (erlend-aasland) [inline ISOLATE_STDLIB_MODULES.rst]:
```suggestion
The following process should be added to the Devguide, and remain until
all modules are converted.
```

File: ISOLATE_STDLIB_MODULES.rst [hunk 11]
```diff
@@ -0,0 +1,240 @@
...
+
+Part 1: Preparation
+...................
+
+1. Open a discussion, either on the bug tracker or on Discourse. Involve the
+   module maintainer and/or code owner. Explain the reason and rationale for
+   the changes.
+2. Identify global state performance bottlenecks, if there are such. Create a
+   proof-of-concept implementation and measure the performance impact. `pyperf`
+   is a good tool for benchmarking.
```
Comment (erlend-aasland) [inline ISOLATE_STDLIB_MODULES.rst]:
```suggestion
2. Identify global state performance bottlenecks.
   Create a proof-of-concept implementation, and measure the performance impact.
   `pyperf` is a good tool for benchmarking.
```

File: ISOLATE_STDLIB_MODULES.rst [hunk 12]
```diff
@@ -0,0 +1,240 @@
...
+Note: this is a suggested implementation plan for a complex module, based on
+lessons learned with other modules. Feel free to simplify it for
+smaller modules.
+
+1. Add Argument Clinic where possible; it enables you to easily use the
+   defining class to fetch module state from type methods.
+2. Prepare for module state; establish a module state `struct`, add an instance
+   as a static global variable, and create helper stubs for fetching the module
+   state.
+3. Add relevant global variables to the module state `struct` and modify code
```
Comment (erlend-aasland) [inline ISOLATE_STDLIB_MODULES.rst]:
```suggestion
3. Add relevant global variables to the module state `struct`, and modify code
```

File: ISOLATE_STDLIB_MODULES.rst [hunk 13]
```diff
@@ -0,0 +1,240 @@
...
+   that accesses the global state to use the module state helpers instead. This
+   step may be broken into several PR's.
+4. Where necessary, convert heap types to static types.
+5. Convert the global module state struct to true module state.
+6. Implement multi-phase initialisation.
+
+Preferably, steps 4 through 6 should all land early in a single alpha
+development phase.
```
Comment (erlend-aasland) [inline ISOLATE_STDLIB_MODULES.rst]:
```suggestion
Steps 4 through 6 should preferably land in a single alpha development phase.
```

File: ISOLATE_STDLIB_MODULES.rst [hunk 14]
```diff
@@ -0,0 +1,240 @@
...
+
+Much of this proposal has already been implemented, partly to solve related
+issues and partly because of misunderstandings and lack of coordination.
+
+We submit this PEP to explain the changes, seek consensus on
+whether they are good,  propose changing modules that do not exhibit
+issues, and set best practices for new modules.
+
+We are aware that the pre-existing changes might need to be reverted if issues
+are found.
```
Comment (erlend-aasland) [inline ISOLATE_STDLIB_MODULES.rst]: The first heap types were added in Python 3.2. Some of these changes are _very_ old. If one concludes there is no gain in converting new types (net win = 0), then there is also no gain in reverting very old changes; IMO it would just be needless code churn, which would be very counter productive (risk of introducing new regressions, require a _lot_ of reviewer resources).

There is already a section further down that says if issues are found, they must be resolved. That should suffice. I suggest dropping this sentence.

Comment (erlend-aasland) [reply ISOLATE_STDLIB_MODULES.rst]: > There is already a section further down that says if issues are found, they must be resolved.

Issues could be general (the overall process) or specific (for example module or type specific). One way of resolving a module specific issue could be to revert changes for that specific module; that can be resolved on a bug ticket; we don't need it in the PEP.

File: ISOLATE_STDLIB_MODULES.rst (hunk 4 above)
Comment (encukou) [reply ISOLATE_STDLIB_MODULES.rst]: Do you have a suggestion?

File: ISOLATE_STDLIB_MODULES.rst (hunk 1 above)
Comment (encukou) [reply ISOLATE_STDLIB_MODULES.rst]: Right. It makes sense to have the testing info in the list of considerations, instead of a new section.

File: ISOLATE_STDLIB_MODULES.rst (hunk 3 above)
Comment (encukou) [reply ISOLATE_STDLIB_MODULES.rst]: Yeah, the title should have XXX as well. I guess “Do not isolate modules in the stdlib” works best, because it needs the other changes.

File: ISOLATE_STDLIB_MODULES.rst [hunk 15]
```diff
@@ -0,0 +1,239 @@
...
+
+See XXX above.
+
+Types that do not need module state access
+------------------------------------------
+
+If Python gets a per-interpreter GIL, these types will need to either
+
+- be converted to heap types, or
+- become immortal (see :pep:`683`).
```
Comment (encukou) [inline ISOLATE_STDLIB_MODULES.rst]: This is covered in the pre-interpreter GIL PEP, so I'll remove it.

File: ISOLATE_STDLIB_MODULES.rst [hunk 16]
```diff
@@ -191,18 +186,25 @@ Beginners should not be affected.
 Reference Implementation
 ========================
 
-Main branch, particularly commits for these issues:
+Most of the changes are now in the main branch, as commits for these issues:
+
+- `bpo-40077, Convert static types to heap types: use PyType_FromSpec() <https://bugs.python.org/issue40077>`_
+- `bpo-40077, Clear static types in Py_Finalize() for embedded Python <https://bugs.python.org/issue46417>`_
```
Comment (erlend-aasland) [inline ISOLATE_STDLIB_MODULES.rst]:
```suggestion
- `bpo-46417, Clear static types in Py_Finalize() for embedded Python <https://bugs.python.org/issue46417>`_
```

File: ISOLATE_STDLIB_MODULES.rst (hunk 4 above)
Comment (erlend-aasland) [reply ISOLATE_STDLIB_MODULES.rst]: I'll try to come up with something.

File: ISOLATE_STDLIB_MODULES.rst (hunk 3 above)
Comment (erlend-aasland) [reply ISOLATE_STDLIB_MODULES.rst]: We should add a short sentence about why having stdlib extension modules _without_ multi-init support and _with_ global state is a problem; for example, such modules would hamper a lot of Eric's work-in-progress.

File: ISOLATE_STDLIB_MODULES.rst [hunk 17]
```diff
@@ -0,0 +1,241 @@
...
+Do not isolate modules in the standard library
+----------------------------------------------
+
+XXX Someone write something here please
+
+
+Open Issues
+===========
+
+See XXX above.
```
Comment (erlend-aasland) [inline ISOLATE_STDLIB_MODULES.rst]: This is already covered in the heap types section, right? I guess we could remove this then.

File: ISOLATE_STDLIB_MODULES.rst (hunk 4 above)
Comment (erlend-aasland) [reply ISOLATE_STDLIB_MODULES.rst]: Re-reading this, I do not see what new value or information it adds. The two first sentences under this sub-heading already explains what the reader needs to know. Here is my suggestion:

```suggestion
```

File: ISOLATE_STDLIB_MODULES.rst [hunk 18]
```diff
@@ -0,0 +1,241 @@
...
+- `GH-26351, Make heap types converted during 3.10 alpha immutable <https://github.com/python/cpython/pull/26351>`_
+
+
+Rejected Ideas
+==============
+
+Do not isolate modules in the standard library
+----------------------------------------------
+
+XXX Someone write something here please
```
Comment (erlend-aasland) [inline ISOLATE_STDLIB_MODULES.rst]: IMO we can reference Eric's work-in-progress (per-interpreter GIL), subinterpreters issues (multiple subinterpreters + static global state => 💥), ... anything else?

**Comment by encukou** [comment]
PR for the two general sections marked “XXX move this to PEP 630”: https://github.com/python/peps/pull/2450

File: ISOLATE_STDLIB_MODULES.rst (hunk 18 above)
Comment (encukou) [reply ISOLATE_STDLIB_MODULES.rst]: Do we want to make this PEP depend on Eric's PEP?
IMO it would be better if it can stand on its own.

Comment (erlend-aasland) [reply ISOLATE_STDLIB_MODULES.rst]: > IMO it would be better if it can stand on its own.

Yes, that's preferable.

Comment (encukou) [reply ISOLATE_STDLIB_MODULES.rst]: So, let's remove the Rejected Ideas section altogether and let the discussion generate some.

Are you happy with your name on this? I think it's ready to post to python-dev.

File: ISOLATE_STDLIB_MODULES.rst [hunk 19]
```diff
@@ -0,0 +1,206 @@
...
+initialization (:pep:`489`) and where possible, all state will be
+stored on module objects rather than in process-global variables.
+
+
+Note on Backdating
+==================
+
+Much of this proposal has already been implemented, partly to solve related
+issues and partly because of misunderstandings and lack of coordination.
+
```
Comment (erlend-aasland) [inline ISOLATE_STDLIB_MODULES.rst]: One last nitpick: IMO, the important piece of information for the reader is that much of this has already been implemented. There's probably a lot of reasons why (finding all of them would be a time-consuming task), but I'm not sure those reasons are valuable information in this context. Suggesting a simplification:

```suggestion
Much of this proposal has already been implemented.
```

**Review by erlend-aasland — COMMENTED** [review]
Your https://github.com/encukou/abi3/pull/24#discussion_r833308377 sounds good to me. I'm fine with putting my name on this (see one last nit); let's post it 🚀

File: ISOLATE_STDLIB_MODULES.rst (hunk 18 above)
Comment (erlend-aasland) [reply ISOLATE_STDLIB_MODULES.rst]: Sounds good to me!

**Comment by erlend-aasland** [comment]
Is there any issues that needs to be resolved before posting to python-dev?

**Comment by encukou** [comment]
Just need more time in the day...

**Comment by encukou** [comment]
PEP PR: https://github.com/python/peps/pull/2499
I'll close this, we won't need an out-of-date copy in this repo.
